// devextreme-angular/esm5/ui/nested/change-dxi.js
/*!
 * devextreme-angular
 * Version: 21.2.1-alpha-21124-0324
 * Build date: Tue May 04 2021
 *
 * Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
import * as tslib_1 from "tslib";
/* tslint:disable:max-line-length */
/* tslint:disable:use-input-property-decorator */
import { Component, NgModule, Host, SkipSelf } from '@angular/core';
import { NestedOptionHost, } from 'devextreme-angular/core';
import { DxiDataChange } from './base/data-change-dxi';
var DxiChangeComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DxiChangeComponent, _super);
    function DxiChangeComponent(parentOptionHost, optionHost) {
        var _this = _super.call(this) || this;
        parentOptionHost.setNestedOption(_this);
        optionHost.setHost(_this, _this._fullOptionPath.bind(_this));
        return _this;
    }
    Object.defineProperty(DxiChangeComponent.prototype, "_optionPath", {
        get: function () {
            return 'changes';
        },
        enumerable: true,
        configurable: true
    });
    DxiChangeComponent.prototype.ngOnDestroy = function () {
        this._deleteRemovedOptions(this._fullOptionPath());
    };
    DxiChangeComponent.ctorParameters = function () { return [
        { type: NestedOptionHost, decorators: [{ type: SkipSelf }, { type: Host }] },
        { type: NestedOptionHost, decorators: [{ type: Host }] }
    ]; };
    DxiChangeComponent = tslib_1.__decorate([
        Component({
            selector: 'dxi-change',
            template: '',
            providers: [NestedOptionHost],
            inputs: [
                'data',
                'index',
                'key',
                'pageIndex',
                'type'
            ],
            styles: ['']
        }),
        tslib_1.__param(0, SkipSelf()), tslib_1.__param(0, Host()),
        tslib_1.__param(1, Host()),
        tslib_1.__metadata("design:paramtypes", [NestedOptionHost,
            NestedOptionHost])
    ], DxiChangeComponent);
    return DxiChangeComponent;
}(DxiDataChange));
export { DxiChangeComponent };
var DxiChangeModule = /** @class */ (function () {
    function DxiChangeModule() {
    }
    DxiChangeModule = tslib_1.__decorate([
        NgModule({
            declarations: [
                DxiChangeComponent
            ],
            exports: [
                DxiChangeComponent
            ],
        })
    ], DxiChangeModule);
    return DxiChangeModule;
}());
export { DxiChangeModule };
