// devextreme-angular/esm5/ui/nested/from.js
/*!
 * devextreme-angular
 * Version: 21.2.1-alpha-21212-0323
 * Build date: Sat Jul 31 2021
 *
 * Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
import * as tslib_1 from "tslib";
/* tslint:disable:max-line-length */
/* tslint:disable:use-input-property-decorator */
import { Component, NgModule, Host, SkipSelf } from '@angular/core';
import { NestedOptionHost, } from 'devextreme-angular/core';
import { DxoAnimationState } from './base/animation-state';
var DxoFromComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DxoFromComponent, _super);
    function DxoFromComponent(parentOptionHost, optionHost) {
        var _this = _super.call(this) || this;
        parentOptionHost.setNestedOption(_this);
        optionHost.setHost(_this, _this._fullOptionPath.bind(_this));
        return _this;
    }
    Object.defineProperty(DxoFromComponent.prototype, "_optionPath", {
        get: function () {
            return 'from';
        },
        enumerable: true,
        configurable: true
    });
    DxoFromComponent.prototype.ngOnInit = function () {
        this._addRecreatedComponent();
    };
    DxoFromComponent.prototype.ngOnDestroy = function () {
        this._addRemovedOption(this._getOptionPath());
    };
    DxoFromComponent.ctorParameters = function () { return [
        { type: NestedOptionHost, decorators: [{ type: SkipSelf }, { type: Host }] },
        { type: NestedOptionHost, decorators: [{ type: Host }] }
    ]; };
    DxoFromComponent = tslib_1.__decorate([
        Component({
            selector: 'dxo-from',
            template: '',
            providers: [NestedOptionHost],
            inputs: [
                'left',
                'opacity',
                'position',
                'scale',
                'top'
            ],
            styles: ['']
        }),
        tslib_1.__param(0, SkipSelf()), tslib_1.__param(0, Host()),
        tslib_1.__param(1, Host()),
        tslib_1.__metadata("design:paramtypes", [NestedOptionHost,
            NestedOptionHost])
    ], DxoFromComponent);
    return DxoFromComponent;
}(DxoAnimationState));
export { DxoFromComponent };
var DxoFromModule = /** @class */ (function () {
    function DxoFromModule() {
    }
    DxoFromModule = tslib_1.__decorate([
        NgModule({
            declarations: [
                DxoFromComponent
            ],
            exports: [
                DxoFromComponent
            ],
        })
    ], DxoFromModule);
    return DxoFromModule;
}());
export { DxoFromModule };
