// devextreme-angular/esm5/ui/nested/notifications.js
/*!
 * devextreme-angular
 * Version: 21.2.1-alpha-21212-0323
 * Build date: Sat Jul 31 2021
 *
 * Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
import * as tslib_1 from "tslib";
/* tslint:disable:max-line-length */
import { Component, NgModule, Host, SkipSelf, Input } from '@angular/core';
import { NestedOptionHost, } from 'devextreme-angular/core';
import { NestedOption } from 'devextreme-angular/core';
var DxoNotificationsComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DxoNotificationsComponent, _super);
    function DxoNotificationsComponent(parentOptionHost, optionHost) {
        var _this = _super.call(this) || this;
        parentOptionHost.setNestedOption(_this);
        optionHost.setHost(_this, _this._fullOptionPath.bind(_this));
        return _this;
    }
    Object.defineProperty(DxoNotificationsComponent.prototype, "showPanel", {
        get: function () {
            return this._getOption('showPanel');
        },
        set: function (value) {
            this._setOption('showPanel', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoNotificationsComponent.prototype, "showPopup", {
        get: function () {
            return this._getOption('showPopup');
        },
        set: function (value) {
            this._setOption('showPopup', value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DxoNotificationsComponent.prototype, "_optionPath", {
        get: function () {
            return 'notifications';
        },
        enumerable: true,
        configurable: true
    });
    DxoNotificationsComponent.prototype.ngOnInit = function () {
        this._addRecreatedComponent();
    };
    DxoNotificationsComponent.prototype.ngOnDestroy = function () {
        this._addRemovedOption(this._getOptionPath());
    };
    DxoNotificationsComponent.ctorParameters = function () { return [
        { type: NestedOptionHost, decorators: [{ type: SkipSelf }, { type: Host }] },
        { type: NestedOptionHost, decorators: [{ type: Host }] }
    ]; };
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], DxoNotificationsComponent.prototype, "showPanel", null);
    tslib_1.__decorate([
        Input(),
        tslib_1.__metadata("design:type", Boolean),
        tslib_1.__metadata("design:paramtypes", [Boolean])
    ], DxoNotificationsComponent.prototype, "showPopup", null);
    DxoNotificationsComponent = tslib_1.__decorate([
        Component({
            selector: 'dxo-notifications',
            template: '',
            providers: [NestedOptionHost],
            styles: ['']
        }),
        tslib_1.__param(0, SkipSelf()), tslib_1.__param(0, Host()),
        tslib_1.__param(1, Host()),
        tslib_1.__metadata("design:paramtypes", [NestedOptionHost,
            NestedOptionHost])
    ], DxoNotificationsComponent);
    return DxoNotificationsComponent;
}(NestedOption));
export { DxoNotificationsComponent };
var DxoNotificationsModule = /** @class */ (function () {
    function DxoNotificationsModule() {
    }
    DxoNotificationsModule = tslib_1.__decorate([
        NgModule({
            declarations: [
                DxoNotificationsComponent
            ],
            exports: [
                DxoNotificationsComponent
            ],
        })
    ], DxoNotificationsModule);
    return DxoNotificationsModule;
}());
export { DxoNotificationsModule };
