// devextreme-angular/esm5/ui/nested/table-resizing.js
/*!
 * devextreme-angular
 * Version: 21.2.1-alpha-21212-0323
 * Build date: Sat Jul 31 2021
 *
 * Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-angular
 */
import * as tslib_1 from "tslib";
/* tslint:disable:max-line-length */
/* tslint:disable:use-input-property-decorator */
import { Component, NgModule, Host, SkipSelf } from '@angular/core';
import { NestedOptionHost, } from 'devextreme-angular/core';
import { DxoHtmlEditorTableResizing } from './base/html-editor-table-resizing';
var DxoTableResizingComponent = /** @class */ (function (_super) {
    tslib_1.__extends(DxoTableResizingComponent, _super);
    function DxoTableResizingComponent(parentOptionHost, optionHost) {
        var _this = _super.call(this) || this;
        parentOptionHost.setNestedOption(_this);
        optionHost.setHost(_this, _this._fullOptionPath.bind(_this));
        return _this;
    }
    Object.defineProperty(DxoTableResizingComponent.prototype, "_optionPath", {
        get: function () {
            return 'tableResizing';
        },
        enumerable: true,
        configurable: true
    });
    DxoTableResizingComponent.prototype.ngOnInit = function () {
        this._addRecreatedComponent();
    };
    DxoTableResizingComponent.prototype.ngOnDestroy = function () {
        this._addRemovedOption(this._getOptionPath());
    };
    DxoTableResizingComponent.ctorParameters = function () { return [
        { type: NestedOptionHost, decorators: [{ type: SkipSelf }, { type: Host }] },
        { type: NestedOptionHost, decorators: [{ type: Host }] }
    ]; };
    DxoTableResizingComponent = tslib_1.__decorate([
        Component({
            selector: 'dxo-table-resizing',
            template: '',
            providers: [NestedOptionHost],
            inputs: [
                'enabled',
                'minColumnWidth',
                'minRowHeight'
            ],
            styles: ['']
        }),
        tslib_1.__param(0, SkipSelf()), tslib_1.__param(0, Host()),
        tslib_1.__param(1, Host()),
        tslib_1.__metadata("design:paramtypes", [NestedOptionHost,
            NestedOptionHost])
    ], DxoTableResizingComponent);
    return DxoTableResizingComponent;
}(DxoHtmlEditorTableResizing));
export { DxoTableResizingComponent };
var DxoTableResizingModule = /** @class */ (function () {
    function DxoTableResizingModule() {
    }
    DxoTableResizingModule = tslib_1.__decorate([
        NgModule({
            declarations: [
                DxoTableResizingComponent
            ],
            exports: [
                DxoTableResizingComponent
            ],
        })
    ], DxoTableResizingModule);
    return DxoTableResizingModule;
}());
export { DxoTableResizingModule };
