// devextreme/renovation/ui/scheduler/workspaces/timeline/header_panel/layout.j.js
/**
 * DevExtreme (renovation/ui/scheduler/workspaces/timeline/header_panel/layout.j.js)
 * Version: 21.1.2 (build 21124-0324)
 * Build date: Tue May 04 2021
 *
 * Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
exports.default = void 0;
var _component_registrator = _interopRequireDefault(require("../../../../../../core/component_registrator"));
var _component = _interopRequireDefault(require("../../../../../component_wrapper/component"));
var _layout = require("./layout");

function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    }
}

function _defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) {
            descriptor.writable = true
        }
        Object.defineProperty(target, descriptor.key, descriptor)
    }
}

function _createClass(Constructor, protoProps, staticProps) {
    if (protoProps) {
        _defineProperties(Constructor.prototype, protoProps)
    }
    if (staticProps) {
        _defineProperties(Constructor, staticProps)
    }
    return Constructor
}

function _inheritsLoose(subClass, superClass) {
    subClass.prototype = Object.create(superClass.prototype);
    subClass.prototype.constructor = subClass;
    _setPrototypeOf(subClass, superClass)
}

function _setPrototypeOf(o, p) {
    _setPrototypeOf = Object.setPrototypeOf || function(o, p) {
        o.__proto__ = p;
        return o
    };
    return _setPrototypeOf(o, p)
}
var TimelineHeaderPanelLayout = function(_BaseComponent) {
    _inheritsLoose(TimelineHeaderPanelLayout, _BaseComponent);

    function TimelineHeaderPanelLayout() {
        return _BaseComponent.apply(this, arguments) || this
    }
    var _proto = TimelineHeaderPanelLayout.prototype;
    _proto.getProps = function() {
        var props = _BaseComponent.prototype.getProps.call(this);
        props.dateCellTemplate = this._createTemplateComponent(props, props.dateCellTemplate);
        props.timeCellTemplate = this._createTemplateComponent(props, props.timeCellTemplate);
        props.dateHeaderTemplate = this._createTemplateComponent(props, props.dateHeaderTemplate);
        props.resourceCellTemplate = this._createTemplateComponent(props, props.resourceCellTemplate);
        return props
    };
    _createClass(TimelineHeaderPanelLayout, [{
        key: "_propsInfo",
        get: function() {
            return {
                twoWay: [],
                allowNull: [],
                elements: [],
                templates: ["dateCellTemplate", "timeCellTemplate", "dateHeaderTemplate", "resourceCellTemplate"],
                props: ["dateHeaderMap", "isRenderDateHeader", "groupPanelCellBaseColSpan", "dateCellTemplate", "timeCellTemplate", "dateHeaderTemplate", "groups", "groupOrientation", "groupByDate", "height", "baseColSpan", "columnCountPerGroup", "className", "resourceCellTemplate"]
            }
        }
    }, {
        key: "_viewComponent",
        get: function() {
            return _layout.TimelineHeaderPanelLayout
        }
    }]);
    return TimelineHeaderPanelLayout
}(_component.default);
exports.default = TimelineHeaderPanelLayout;
(0, _component_registrator.default)("dxTimelineHeaderPanelLayout", TimelineHeaderPanelLayout);
module.exports = exports.default;
module.exports.default = exports.default;
