// devextreme/esm/renovation/ui/scheduler/workspaces/base/virtual-cell.js
/**
 * DevExtreme (esm/renovation/ui/scheduler/workspaces/base/virtual-cell.js)
 * Version: 21.1.2 (build 21124-0324)
 * Build date: Tue May 04 2021
 *
 * Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
import {
    createVNode
} from "inferno";
import {
    BaseInfernoComponent,
    normalizeStyles
} from "@devextreme/vdom";
import {
    addWidthToStyle
} from "../utils";
export var viewFunction = viewModel => createVNode(1, "td", "dx-scheduler-virtual-cell", null, 1, {
    style: normalizeStyles(viewModel.style)
});
export var VirtualCellProps = {
    width: 0
};
export class VirtualCell extends BaseInfernoComponent {
    constructor(props) {
        super(props);
        this.state = {}
    }
    get style() {
        var {
            width: width
        } = this.props;
        var {
            style: style
        } = this.restAttributes;
        return addWidthToStyle(width, style)
    }
    get restAttributes() {
        var _this$props = this.props,
            restProps = _objectWithoutPropertiesLoose(_this$props, ["width"]);
        return restProps
    }
    render() {
        var props = this.props;
        return viewFunction({
            props: _extends({}, props),
            style: this.style,
            restAttributes: this.restAttributes
        })
    }
}
VirtualCell.defaultProps = _extends({}, VirtualCellProps);
