// devextreme-vue/core/extension-component.js
/*!
 * devextreme-vue
 * Version: 21.2.1-alpha-21212-0323
 * Build date: Sat Jul 31 2021
 *
 * Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file in the root of the project for details.
 *
 * https://github.com/DevExpress/devextreme-vue
 */

"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.initDxExtensionComponent = void 0;
var vue_1 = require("vue");
var component_1 = require("./component");
var vue_helper_1 = require("./vue-helper");
function initDxExtensionComponent() {
    return vue_1.defineComponent({
        extends: component_1.initBaseComponent(),
        created: function () {
            var nodeOptions = vue_helper_1.getNodeOptions(this);
            nodeOptions.$_isExtension = true;
            nodeOptions.$_attachTo = this.attachTo.bind(this);
        },
        mounted: function () {
            this.$el.setAttribute("isExtension", "true");
            var componentOptions = vue_helper_1.getNodeOptions(this);
            if (componentOptions && componentOptions.$_hasOwner) {
                return;
            }
            this.attachTo(this.$el);
        },
        methods: {
            attachTo: function (element) {
                this.$_createWidget(element);
            }
        }
    });
}
exports.initDxExtensionComponent = initDxExtensionComponent;
