// devextreme/esm/renovation/ui/scroll_view/scrollable.j.js
/**
 * DevExtreme (esm/renovation/ui/scroll_view/scrollable.j.js)
 * Version: 21.1.2 (build 21124-0324)
 * Build date: Tue May 04 2021
 *
 * Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import registerComponent from "../../../core/component_registrator";
import BaseComponent from "../../component_wrapper/component";
import {
    Scrollable as ScrollableComponent
} from "./scrollable";
export default class Scrollable extends BaseComponent {
    content() {
        return this._toPublicElement(this.viewRef.content())
    }
    scrollBy(distance) {
        return this.viewRef.scrollBy(distance)
    }
    update() {
        return this.viewRef.update()
    }
    release() {
        return this.viewRef.release()
    }
    refresh() {
        return this.viewRef.refresh()
    }
    scrollTo(targetLocation) {
        return this.viewRef.scrollTo(targetLocation)
    }
    scrollToElement(element) {
        return this.viewRef.scrollToElement(this._patchElementParam(element))
    }
    scrollHeight() {
        return this.viewRef.scrollHeight()
    }
    scrollWidth() {
        return this.viewRef.scrollWidth()
    }
    scrollOffset() {
        return this.viewRef.scrollOffset()
    }
    scrollTop() {
        return this.viewRef.scrollTop()
    }
    scrollLeft() {
        return this.viewRef.scrollLeft()
    }
    clientHeight() {
        return this.viewRef.clientHeight()
    }
    clientWidth() {
        return this.viewRef.clientWidth()
    }
    getScrollElementPosition(element, direction) {
        return this.viewRef.getScrollElementPosition(this._patchElementParam(element), direction)
    }
    _getActionConfigs() {
        return {
            onScroll: {},
            onUpdated: {},
            onPullDown: {},
            onReachBottom: {},
            onStart: {},
            onEnd: {},
            onBounce: {},
            onStop: {}
        }
    }
    get _propsInfo() {
        return {
            twoWay: [],
            allowNull: [],
            elements: [],
            templates: [],
            props: ["useNative", "direction", "showScrollbar", "bounceEnabled", "scrollByContent", "scrollByThumb", "updateManually", "classes", "pullDownEnabled", "reachBottomEnabled", "forceGeneratePockets", "needScrollViewContentWrapper", "needScrollViewLoadPanel", "onScroll", "onUpdated", "onPullDown", "onReachBottom", "pullingDownText", "pulledDownText", "refreshingText", "reachBottomText", "aria", "disabled", "height", "rtlEnabled", "visible", "width", "useSimulatedScrollbar", "inertiaEnabled", "useKeyboard", "onStart", "onEnd", "onBounce", "onStop"]
        }
    }
    get _viewComponent() {
        return ScrollableComponent
    }
}
registerComponent("dxScrollable", Scrollable);
