// html2canvas/dist/lib/dom/element-container.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElementContainer = void 0;
var index_1 = require("../css/index");
var bounds_1 = require("../css/layout/bounds");
var node_parser_1 = require("./node-parser");
var ElementContainer = /** @class */ (function () {
    function ElementContainer(context, element) {
        this.context = context;
        this.styles = new index_1.CSSParsedDeclaration(context, window.getComputedStyle(element, null));
        this.textNodes = [];
        this.elements = [];
        if (node_parser_1.isHTMLElementNode(element)) {
            if (this.styles.animationDuration > 0) {
                element.style.animationDuration = '0s';
            }
            if (this.styles.transitionDuration > 0) {
                element.style.transitionDuration = '0s';
            }
            if (this.styles.transform !== null) {
                // getBoundingClientRect takes transforms into account
                element.style.transform = 'none';
            }
        }
        this.bounds = bounds_1.parseBounds(this.context, element);
        this.flags = 0;
    }
    return ElementContainer;
}());
exports.ElementContainer = ElementContainer;
