// devextreme/esm/renovation/ui/scheduler/workspaces/timeline/header_panel/layout.j.js
/**
 * DevExtreme (esm/renovation/ui/scheduler/workspaces/timeline/header_panel/layout.j.js)
 * Version: 21.1.2 (build 21124-0324)
 * Build date: Tue May 04 2021
 *
 * Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import registerComponent from "../../../../../../core/component_registrator";
import BaseComponent from "../../../../../component_wrapper/component";
import {
    TimelineHeaderPanelLayout as TimelineHeaderPanelLayoutComponent
} from "./layout";
export default class TimelineHeaderPanelLayout extends BaseComponent {
    getProps() {
        var props = super.getProps();
        props.dateCellTemplate = this._createTemplateComponent(props, props.dateCellTemplate);
        props.timeCellTemplate = this._createTemplateComponent(props, props.timeCellTemplate);
        props.dateHeaderTemplate = this._createTemplateComponent(props, props.dateHeaderTemplate);
        props.resourceCellTemplate = this._createTemplateComponent(props, props.resourceCellTemplate);
        return props
    }
    get _propsInfo() {
        return {
            twoWay: [],
            allowNull: [],
            elements: [],
            templates: ["dateCellTemplate", "timeCellTemplate", "dateHeaderTemplate", "resourceCellTemplate"],
            props: ["dateHeaderMap", "isRenderDateHeader", "groupPanelCellBaseColSpan", "dateCellTemplate", "timeCellTemplate", "dateHeaderTemplate", "groups", "groupOrientation", "groupByDate", "height", "baseColSpan", "columnCountPerGroup", "className", "resourceCellTemplate"]
        }
    }
    get _viewComponent() {
        return TimelineHeaderPanelLayoutComponent
    }
}
registerComponent("dxTimelineHeaderPanelLayout", TimelineHeaderPanelLayout);
