// devextreme/esm/renovation/ui/scheduler/workspaces/base/date_table/all_day_panel/title.js
/**
 * DevExtreme (esm/renovation/ui/scheduler/workspaces/base/date_table/all_day_panel/title.js)
 * Version: 21.2.0 (build 21212-0323)
 * Build date: Sat Jul 31 2021
 *
 * Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
import _extends from "@babel/runtime/helpers/esm/extends";
var _excluded = ["className", "isSetTitleClass", "visible"];
import {
    createVNode,
    normalizeProps
} from "inferno";
import {
    InfernoWrapperComponent
} from "@devextreme/vdom";
import messageLocalization from "../../../../../../../localization/message";
import {
    combineClasses
} from "../../../../../../utils/combine_classes";
export var viewFunction = viewModel => normalizeProps(createVNode(1, "div", viewModel.classes, viewModel.text, 0, _extends({}, viewModel.restAttributes)));
export var AllDayPanelTitleProps = {
    className: "",
    visible: true,
    isSetTitleClass: true
};
import {
    createReRenderEffect
} from "@devextreme/vdom";
export class AllDayPanelTitle extends InfernoWrapperComponent {
    constructor(props) {
        super(props);
        this.state = {}
    }
    createEffects() {
        return [createReRenderEffect()]
    }
    get text() {
        return messageLocalization.format("dxScheduler-allDay")
    }
    get classes() {
        return combineClasses({
            "dx-scheduler-all-day-title": this.props.isSetTitleClass,
            "dx-scheduler-all-day-title-hidden": !this.props.visible,
            [this.props.className]: !!this.props.className
        })
    }
    get restAttributes() {
        var _this$props = this.props,
            restProps = _objectWithoutPropertiesLoose(_this$props, _excluded);
        return restProps
    }
    render() {
        var props = this.props;
        return viewFunction({
            props: _extends({}, props),
            text: this.text,
            classes: this.classes,
            restAttributes: this.restAttributes
        })
    }
}
AllDayPanelTitle.defaultProps = _extends({}, AllDayPanelTitleProps);
