// devextreme/esm/renovation/ui/toolbar/toolbar_props.js
/**
 * DevExtreme (esm/renovation/ui/toolbar/toolbar_props.js)
 * Version: 21.2.0 (build 21212-0323)
 * Build date: Sat Jul 31 2021
 *
 * Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import _extends from "@babel/runtime/helpers/esm/extends";
import {
    BaseWidgetProps
} from "../common/base_props";
export var CollectionWidgetItem = {};
export var ToolbarTextBoxProps = {
    value: ""
};
export var ToolbarCheckBoxProps = {
    value: false
};
export var ToolbarButtonGroupItemProps = _extends({}, CollectionWidgetItem);
export var ToolbarButtonGroupItemPropsType = {};
export var ToolbarButtonGroupProps = {};
export var ToolbarButtonProps = {};
export var ToolbarDropDownButtonItemProps = _extends({}, CollectionWidgetItem);
export var ToolbarDropDownButtonItemPropsType = {};
export var ToolbarDropDownButtonProps = {};
export var ToolbarItem = _extends({}, CollectionWidgetItem);
export var ToolbarItemType = {};
export var ToolbarProps = _extends({}, BaseWidgetProps);
