// devextreme/cjs/renovation/ui/scroll_view/scrollable.j.js
/**
 * DevExtreme (cjs/renovation/ui/scroll_view/scrollable.j.js)
 * Version: 21.1.2 (build 21124-0324)
 * Build date: Tue May 04 2021
 *
 * Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
"use strict";
exports.default = void 0;
var _component_registrator = _interopRequireDefault(require("../../../core/component_registrator"));
var _component = _interopRequireDefault(require("../../component_wrapper/component"));
var _scrollable = require("./scrollable");

function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    }
}

function _defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) {
            descriptor.writable = true
        }
        Object.defineProperty(target, descriptor.key, descriptor)
    }
}

function _createClass(Constructor, protoProps, staticProps) {
    if (protoProps) {
        _defineProperties(Constructor.prototype, protoProps)
    }
    if (staticProps) {
        _defineProperties(Constructor, staticProps)
    }
    return Constructor
}

function _inheritsLoose(subClass, superClass) {
    subClass.prototype = Object.create(superClass.prototype);
    subClass.prototype.constructor = subClass;
    _setPrototypeOf(subClass, superClass)
}

function _setPrototypeOf(o, p) {
    _setPrototypeOf = Object.setPrototypeOf || function(o, p) {
        o.__proto__ = p;
        return o
    };
    return _setPrototypeOf(o, p)
}
var Scrollable = function(_BaseComponent) {
    _inheritsLoose(Scrollable, _BaseComponent);

    function Scrollable() {
        return _BaseComponent.apply(this, arguments) || this
    }
    var _proto = Scrollable.prototype;
    _proto.content = function() {
        return this._toPublicElement(this.viewRef.content())
    };
    _proto.scrollBy = function(distance) {
        return this.viewRef.scrollBy(distance)
    };
    _proto.update = function() {
        return this.viewRef.update()
    };
    _proto.release = function() {
        return this.viewRef.release()
    };
    _proto.refresh = function() {
        return this.viewRef.refresh()
    };
    _proto.scrollTo = function(targetLocation) {
        return this.viewRef.scrollTo(targetLocation)
    };
    _proto.scrollToElement = function(element) {
        return this.viewRef.scrollToElement(this._patchElementParam(element))
    };
    _proto.scrollHeight = function() {
        return this.viewRef.scrollHeight()
    };
    _proto.scrollWidth = function() {
        return this.viewRef.scrollWidth()
    };
    _proto.scrollOffset = function() {
        return this.viewRef.scrollOffset()
    };
    _proto.scrollTop = function() {
        return this.viewRef.scrollTop()
    };
    _proto.scrollLeft = function() {
        return this.viewRef.scrollLeft()
    };
    _proto.clientHeight = function() {
        return this.viewRef.clientHeight()
    };
    _proto.clientWidth = function() {
        return this.viewRef.clientWidth()
    };
    _proto.getScrollElementPosition = function(element, direction) {
        return this.viewRef.getScrollElementPosition(this._patchElementParam(element), direction)
    };
    _proto._getActionConfigs = function() {
        return {
            onScroll: {},
            onUpdated: {},
            onPullDown: {},
            onReachBottom: {},
            onStart: {},
            onEnd: {},
            onBounce: {},
            onStop: {}
        }
    };
    _createClass(Scrollable, [{
        key: "_propsInfo",
        get: function() {
            return {
                twoWay: [],
                allowNull: [],
                elements: [],
                templates: [],
                props: ["useNative", "direction", "showScrollbar", "bounceEnabled", "scrollByContent", "scrollByThumb", "updateManually", "classes", "pullDownEnabled", "reachBottomEnabled", "forceGeneratePockets", "needScrollViewContentWrapper", "needScrollViewLoadPanel", "onScroll", "onUpdated", "onPullDown", "onReachBottom", "pullingDownText", "pulledDownText", "refreshingText", "reachBottomText", "aria", "disabled", "height", "rtlEnabled", "visible", "width", "useSimulatedScrollbar", "inertiaEnabled", "useKeyboard", "onStart", "onEnd", "onBounce", "onStop"]
            }
        }
    }, {
        key: "_viewComponent",
        get: function() {
            return _scrollable.Scrollable
        }
    }]);
    return Scrollable
}(_component.default);
exports.default = Scrollable;
(0, _component_registrator.default)("dxScrollable", Scrollable);
module.exports = exports.default;
module.exports.default = exports.default;
