// devextreme/esm/renovation/ui/scheduler/appointment_tooltip/item_content.js
/**
 * DevExtreme (esm/renovation/ui/scheduler/appointment_tooltip/item_content.js)
 * Version: 21.1.2 (build 21124-0324)
 * Build date: Tue May 04 2021
 *
 * Copyright (c) 2012 - 2021 Developer Express Inc. ALL RIGHTS RESERVED
 * Read about DevExtreme licensing here: https://js.devexpress.com/Licensing/
 */
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
import _extends from "@babel/runtime/helpers/esm/extends";
import {
    createVNode,
    normalizeProps
} from "inferno";
import {
    BaseInfernoComponent
} from "@devextreme/vdom";
export var viewFunction = viewModel => normalizeProps(createVNode(1, "div", "dx-tooltip-appointment-item-content ".concat(viewModel.props.className), [createVNode(1, "div", "dx-tooltip-appointment-item-content-subject", viewModel.props.text, 0), createVNode(1, "div", "dx-tooltip-appointment-item-content-date", viewModel.props.formattedDate, 0)], 4, _extends({}, viewModel.restAttributes)));
export var TooltipItemContentProps = {
    className: "",
    text: "",
    formattedDate: ""
};
export class TooltipItemContent extends BaseInfernoComponent {
    constructor(props) {
        super(props);
        this.state = {}
    }
    get restAttributes() {
        var _this$props = this.props,
            restProps = _objectWithoutPropertiesLoose(_this$props, ["className", "formattedDate", "text"]);
        return restProps
    }
    render() {
        var props = this.props;
        return viewFunction({
            props: _extends({}, props),
            restAttributes: this.restAttributes
        })
    }
}
TooltipItemContent.defaultProps = _extends({}, TooltipItemContentProps);
